agora/Agora/Effect/TreasuryWithdrawal.hs

```diff
@@ -148,8 +139,17 @@ treasuryWithdrawalValidator currSymbol = makeEffect currSymbol $
                   effInput.address #== pfield @"address" # pfromData x
               )
             # pfromData txInfo.outputs
+        inputsAreOnlyTreasuries =
+          pall
+            # plam
+              ( \((pfield @"_0" #) . pfromData -> cred) ->
+                  cred #== pfield @"credential" # effInput.address
+                    #|| pelem # cred # datum.treasuries
+              )
+            # inputValues
```

Comment:
Okay, I believe this almost is correct. But I'm unsure if it's possible to create a valid transaction as a 3rd party now.

e.g. your average transaction will actually probably have a signer also spend some of their own ADA in order to pay the fee. Furthermore, the collateral also needs to be able to come from a 3rd party address. I guess these would be `PubKeyHash`. So maybe you can _allow_ all `PubKeyCredential`s, because they will never have logic for GAT burning anyway (they are users, essentially!)

Reply:
It specifically allows the signer to have minimum Ada amount as it excludes the some amount from the input treasury sum and output treasury sum.
I specifically made it to do that because it can limit the input for specific scripts and users. That being said, I did not know about the collateral participation. Can you explain more about such situations?

Reply:
Please read https://docs.cardano.org/plutus/collateral-mechanism, and if desired, https://hydra.iohk.io/build/7172824/download/1/alonzo-changes.pdf (specifically section 4).

The alonzo.cddl file also shows where the collateral lives in the transaction:
https://github.com/input-output-hk/cardano-ledger/blob/master/eras/alonzo/test-suite/cddl-files/alonzo.cddl#L62

The TL;DR is: Collaterals are _inputs_ that are _from a PubKeyHash_ that must be able to cover a set percentage of the transaction's fees should phase-2 validation fail (Plutus execution). Furthermore, these inputs must hold _only_ ADA. And are fully seized upon failure to validate (thus no other tokens should be present, to pay the validator node!).